### DISCUSSION TITLE
Using arbitrary list of Vectors in epsilon_func

### DISCUSSION THREAD

**Discussion by raksharuia** [NanoComp/meep#2455] [Q&A]
Hello MEEP community,

I am trying to create a custom-geometry using `epsilon_func` in simulation class and have an input list of `Vector3` points.

The list of `Vector3` points is stored in a variable 

`vecs_xyz = [[Vector3<0.2, 0.0, 0.0>, Vector3<0.177, 0.092, 0.0>, .....]` and I want to use them in `epsilon_func`. I tried something like

```python
def custom_geo(vecs_xyz):
    for p in vecs_xyz:
        yield 12 

sim_vec = mp.Simulation(cell_size=mp.Vector3(len(vecs_xyz),len(vecs_xyz),0), 
epsilon_func=custom_geo(vecs_xyz), resolution=50)
```

My guess is that I am supposed to call the custom_geom() function in some loop for it to take Vector3 position from 'vecs_xyz()' one after another (I do need to put some conditions for each point, not included to reduce complexity). 

What would be the best way to achieve this? Any directions are greatly appreciated :)

I tried following this example I found (below) :

The example I found created a simple ring resonator ([https://meep.readthedocs.io/en/latest/Subpixel_Smoothing/#enabling-averaging-for-material-function])(url) for the material function.

I did some modification to work with epsilon_func:
```python
def ring_resonator(p):
    rr = [(p.x**2+p.y**2)**0.5]
    if (rr > rad) and (rr < rad+w):
        return 12 
    return 1 

ring_resonator.do_averaging = False
sim = mp.Simulation(cell_size=mp.Vector3(5,5), epsilon_func=ring_resonator, resolution=100, boundary_layers[mp.PML(0.1)])
```

![image](https://user-images.githubusercontent.com/37671249/229772037-ce7ac1da-121d-4671-844f-eb8a1576dec7.png)

Here, the list of `p` Vectors are auto-generated. However, I have a set of predefined Vector3 positions which I am unable to process directly into the code.

**Comment by smartalecH** [comment]
(I formatted your comment with proper code blocks... it was hard to read before).

The trick with these functions is that the _input_ is an arbitrary point (meep will loop through each point and pass that point into the function). Your task then is to determine what to _do_ with that point. You _could_ try something like this...

```python

vecs_xyz = [[Vector3<0.2, 0.0, 0.0>, Vector3<0.177, 0.092, 0.0>, .....] # pseudocode
def custom_geo(p):
    if p in vecs_xyz:
        return 12
    else:
        return 1

```

This will be problematic though if the meep grid points don't align perfectly with your list of points...

**Reply by raksharuia** [reply]
Thank you @smartalecH. I am sorry for the code blocks. I did type them in the code format but did not verify if everything looked fine. 

I understand a bit more about how these custom functions work. Is there any detailed documentation for these functions - to know how MEEP determines and loops through the arbitrary input 'p' position vector? It is going to be important for me to set my list of points as close as possible to MEEP points.

**Reply by raksharuia** [reply]
@smartalecH - working more into it, I find some weird thing happening with the suggested solution.

My list of vectors comes from a pseudocode and has around 50 points. I could set tolerance and play with resolution (I guess) to have grid points close to my points. However, I see the following with your solution (I included a print statement to see which 'p' -grid points- were found)

Apparently, it doesn't return all the points and I do not understand why it finds the same point multiple times. I think understanding how grid points all looped through and compared to values in the list would then help set better custom structures from me.



```vecs_xyz = [[Vector3<0.2, 0.0, 0.0>, Vector3<0.177, 0.092, 0.0>, .....] # pseudocode```

```
def custom_geo(p):
    if p in vecs_xyz:
        print('I found the point',p)
        return 12
    else:
        return 1
```

**Reply by raksharuia** [reply]
![a](https://user-images.githubusercontent.com/37671249/231171097-e9f79c9c-9df5-4fb3-a615-f518ce2b1eac.png)
